src/RTMClient.ts

```diff
@@ -365,18 +367,40 @@ export class RTMClient extends EventEmitter {
 
     // delegate behavior to state machine
     this.stateMachine.handle('start');
+
+    // return a promise that resolves with the connection information
+    return new Promise((resolve, reject) => {
+      this.once('authenticated', (result) => {
+        this.removeListener('disconnected', reject);
+        resolve(result);
+      });
+      this.once('disconnected', (err) => {
+        this.removeListener('authenticated', resolve);
+        reject(err);
+      });
+    });
   }
 
   /**
    * End an RTM session. After this method is called no messages will be sent or received unless you call
    * start() again later.
    */
-  public disconnect(): void {
-    // TODO: should this return a Promise<void>?
-    this.logger.debug('manual disconnect');
+  public disconnect(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.logger.debug('manual disconnect');
+
+      // resolve (or reject) on disconnect
+      this.once('disconnected', (err) => {
+        if (err instanceof Error) {
+          reject(err);
+        } else {
+          resolve();
+        }
+      });
 
-    // delegate behavior to state machine
-    this.stateMachine.handle('explicit disconnect');
+      // delegate behavior to state machine
+      this.stateMachine.handle('explicit disconnect');
```

Comment:
any reason to kick this event off inside the returned `Promise`'s constructor, as opposed to the line before the return statement?

Reply:
If the web socket immediately errors on closing (that is, on the same tick of the event loop as `handle` is called) then the event handler won't be registered by the time the state machine transitions, and the promise will never reject.

Alternatively, the web socket could also immediately close (e.g. when mocked for testing) and the same problem would happen, though the promise wouldn't *resolve*.

I'm just being cautious that the reactive bit of code is set up before I set off the chain reaction. Don't want any race conditions happening 😅